sql-plugin/src/main/scala/com/nvidia/spark/rapids/AstUtil.scala

```diff
@@ -119,4 +124,169 @@ object AstUtil {
       }
     }
   }
+
+  /**
+   * [NonAstJoinCondSplitStrategy] is to transform original join condition into extra
+   * filter/project when necessary. It's targeted for some cases join condition is not fully
+   * evaluated by ast.
+   * Base on join condition, it can be transformed into three major strategies:
+   * (1) [NoopJoinCondSplit]: noop when join condition can be fully evaluated with ast.
+   * (2) [JoinCondSplitAsPostFilter]: entire join condition is pulled out as a post filter
+   * after join condition.
+   * (3) [JoinCondSplitAsProject]: extract not supported join condition into pre-project nodes
+   * on each join child. One extra project node is introduced to remove intermediate attributes.
+   */
+  abstract class JoinCondSplitStrategy(left: Seq[NamedExpression],
+      right: Seq[NamedExpression], buildSide: GpuBuildSide) extends Serializable {
+
+    // Actual output of build/stream side project due to join condition split
+    private[this] val (buildOutputAttr, streamOutputAttr) = buildSide match {
+      case GpuBuildLeft => (joinLeftOutput, joinRightOutput)
+      case GpuBuildRight => (joinRightOutput, joinLeftOutput)
+    }
+
+    def astCondition(): Option[Expression]
+
+    def processBuildSideAndClose(input: ColumnarBatch): ColumnarBatch = input
+
+    def processStreamSideAndClose(input: ColumnarBatch): ColumnarBatch = input
+
+    def processPostJoin(iter: Iterator[ColumnarBatch]): Iterator[ColumnarBatch] = iter
+
+    def leftOutput(): Seq[NamedExpression] = left
+
+    def rightOutput(): Seq[NamedExpression] = right
+
+    // This is the left side child of join. In `split as project` strategy, it may be different
+    // from original left child with extracted join condition attribute.
+    def joinLeftOutput(): Seq[Attribute] = leftOutput.toSeq.map(expr => expr.toAttribute)
+
+    // This is the right side child of join. In `split as project` strategy, it may be different
+    // from original right child with extracted join condition attribute.
+    def joinRightOutput(): Seq[Attribute] = rightOutput.toSeq.map(expr => expr.toAttribute)
+
+    // Updated build attribute list after join condition split as project node.
+    // It may include extra attributes from split join condition.
+    def buildSideOutput(): scala.collection.Seq[Attribute] = buildOutputAttr
```

Comment:
In general, stylistically you should not necessarily return a `scala.collection.Seq`, you generally should return a `scala.Seq` since the return value is mutable.

Reply:
I c. It's migrated to ```scala.collection.immutable.Seq```. Let me have a try on this.

https://github.com/scala/scala/blob/2.12.x/src/library/scala/package.scala#L48-L49

https://github.com/scala/scala/blob/2.13.x/src/library/scala/package.scala#L58-L60